org.mixedrealitytoolkit.input/InteractionModes/NearInteractionModeDetector.cs

```diff
@@ -16,58 +16,101 @@ namespace MixedReality.Toolkit.Input
         {
             bool result = base.IsModeDetected() || IsNearInteractorSelecting();
             if (result)
             {
-                for (int i = 0; i < previouslyDetectedColliders.Count; i++)
+                UpdateCurrentlyDetectedInteractables();
+                UpdateProximityExited();
+                UpdateProximityEntered();
+            }
+            return result;
+        }
+
+        /// <summary>
+        /// Calls OnProximityExited for all interactables that were previously ProximityHover detected but are no longer detected.
+        /// </summary>
+        private void UpdateProximityExited()
+        {
+            noLongerDetectedInteractables.Clear();
+
+            foreach (IXRProximityInteractable previouslyDetectedInteractable in previouslyDetectedInteractables)
+            {
+                if (!currentlyDetectedInteractables.Contains(previouslyDetectedInteractable))
+                {
+                    noLongerDetectedInteractables.Add(previouslyDetectedInteractable);
+                }
+            }
+
+            foreach (IXRProximityInteractable noLongerDetectedInteractable in noLongerDetectedInteractables)
+            {
+                noLongerDetectedInteractable.OnProximityExited(new ProximityHoverExitedEventArgs(noLongerDetectedInteractable));
+                previouslyDetectedInteractables.Remove(noLongerDetectedInteractable);
+            }
+        }
+
+        /// <summary>
+        /// Call OnProximityEntered for all interactables that are currently detected but were not detected previously.
+        /// </summary>
+        private void UpdateProximityEntered()
+        {
+            foreach (IXRProximityInteractable currentlyDetectedInteractable in currentlyDetectedInteractables)
+            {
+                if (previouslyDetectedInteractables.Add(currentlyDetectedInteractable))
                 {
-                    Collider previouslyDetectedCollider = previouslyDetectedColliders[i];
-                    if (!DetectedColliders.Contains(previouslyDetectedCollider) && previouslyDetectedCollider != null)
-                    {
-                        IXRProximityInteractable nearInteractionMode = previouslyDetectedCollider.GetComponent<IXRProximityInteractable>();
-                        if (nearInteractionMode != null)
-                        {
-                            foreach (XRBaseInteractor xrBaseInteractor in nearInteractors)
-                            {
-                                previouslyDetectedCollider.GetComponent<IXRProximityInteractable>().OnProximityExited(new ProximityExitedEventArgs(this, previouslyDetectedCollider, xrBaseInteractor));
-                            }
-                        }
-                        previouslyDetectedColliders.Remove(previouslyDetectedCollider);
-                    }
+                    currentlyDetectedInteractable.OnProximityEntered(new ProximityHoverEnteredEventArgs(currentlyDetectedInteractable));
```

Comment:
I just realized that you are passing the "interactable" into the args.   I don't see how that helps.  The interactable will always be the target object.   

The args should contain the thing doing the proximity hovering, the interactor or the ProximityDetector

Reply:
`ProximityHoverEnteredEventArgs(this));`

Reply:
I was under the wrong impression that an Interactable would be enough to uniquely identify interaction and got confused with some previous feedback.  I reverted back to use Interactors in commit https://github.com/MixedRealityToolkit/MixedRealityToolkit-Unity/pull/659/commits/21f16fc4597f04b28bbe910fcfd1c1bc6a21adf0
Thank you for the offline explanation on this issue btw.